ddl/ddl_worker.go

```diff
@@ -527,6 +535,42 @@ func (w *worker) handleDDLJobQueue(d *ddlCtx) error {
 	}
 }
 
+func getDDLJobsInQueue(t *meta.Meta, jobListKey meta.JobListKeyType) ([]*model.Job, error) {
+	cnt, err := t.DDLJobQueueLen(jobListKey)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	jobs := make([]*model.Job, cnt)
+	for i := range jobs {
+		jobs[i], err = t.GetDDLJobByIdx(int64(i), jobListKey)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return jobs, nil
+}
+
+func removeJobFromCancellingList(t *meta.Meta, job *model.Job) error {
+	cancelledJobs, err := getDDLJobsInQueue(t, meta.CancelJobListKey)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	found := false
+	pos := 0
+	for i, canceledJob := range cancelledJobs {
+		if canceledJob.ID == job.ID {
+			found = true
+			pos = i
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+	cancelledJobs = append(cancelledJobs[:pos], cancelledJobs[pos+1:]...)
```

Comment:
Don't you store it back?

Reply:
Is storing back required?

Reply:
If not, why assign a value to a local variable at the end of this function?
Every time it checks the cancel queue, it reads from metadata, not memory.

Reply:
This function will be called `finishDDLJob`. In `finishDDLJob`, I think it should be removed from the queue and there is no need to store it back.

Reply:
I see, we need to store it back(the 'metadata' in TiKV) or the persisted queue will be increasing still.

Reply:
When should we remove the job from the `cancelling queue`?

Reply:
We may need to store it back once after we did the check here. otherwise, the rolling back logic will get in a loop.

Reply:
But this function is called in `finishDDLJob` and the job has been rollback. In `finishDDLJob`, do you think we should remove it from the queue? Or when the job should be removed from the queue?